### ISSUE TITLE
Bug: `no-undef` cant show "Cannot access 'x' before initialization" like errors

### ISSUE THREAD

**Issue by Akif9748** [eslint/eslint#16549]
### Environment

Node version: v18.12.1
npm version: v8.18.0
Local ESLint version: v8.27.0
Global ESLint version: v8.25.0 (Currently used)
Operating System: win32 10.0.19045

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```json
{
    "env": {
        "node": true,
        "commonjs": true,
        "es2021": true
    },
    "extends": "eslint:recommended",
    "overrides": [],
    "parserOptions": {
        "ecmaVersion": "latest"
    },
    "rules": {
        "no-unused-vars": "off",
        "no-empty": "off"
    }
}
```
</details>

Eslint can show error:
```js
 prefi; // here
    const guild = await Guild(guildId);
    let { prefix } = guild;
```
```
19:8  error  'prefi' is not defined  no-undef
✖ 1 problem (1 error, 0 warnings)
```
Eslint cant show error:

```js
 `${prefix}` // here
    const guild = await Guild(guildId);
    let { prefix } = guild;
```
But JavaScript gives error:
``` 
ReferenceError: Cannot access 'prefix' before initialization
```
The variable named prefix, not declared other parts the in code. I think eslint need to show this error.

### What did you expect to happen?

An error like this:
```
19:8  error  'prefix' is not defined  no-undef
✖ 1 problem (1 error, 0 warnings)
```
It will be not same.

### What actually happened?

Not eslint output. 0 problem.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I hope I didn't open the issue in the wrong place or it's not stupid...

**Comment by mdjermanovic** [comment]
Hi @Akif9748, thanks for the issue!

`no-undef` works as intended because `prefix` is declared (although not yet initialized) in the scope.

You can use the [no-use-before-define](https://eslint.org/docs/latest/rules/no-use-before-define) rule to get warnings about variables that are used before initialization. Here's a [playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiIGAke3ByZWZpeH1gIC8vIGhlcmVcbiAgICBjb25zdCBndWlsZCA9IGF3YWl0IEd1aWxkKGd1aWxkSWQpO1xuICAgIGxldCB7IHByZWZpeCB9ID0gZ3VpbGQ7Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11c2UtYmVmb3JlLWRlZmluZSI6WyJlcnJvciJdfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==).

**Comment by Akif9748** [comment]
Thanks you so much. I just needed this error, I added it to the config and it works
![image](https://user-images.githubusercontent.com/70021050/201632260-afbca0c9-64ff-44f6-a2e2-b37b458c44f6.png)

**Comment by Slapbox** [comment]
@mdjermanovic why isn't `no-use-before-define` part of the recommended rule set? I assumed it certainly would be (a mistake to assume on my part) but I ended up wasting an hour trying to figure out why ESLint wasn't catching this - expecting it to be a default like `no-undef`.

Secondary thought - although I'm sure this change will not be made - the existence of `no-undef` and `no-undefined` is a bit confusing, especially since `no-undef` really means `no-undeclared`.

**Comment by mdjermanovic** [comment]
> why isn't `no-use-before-define` part of the recommended rule set?

This was discussed in https://github.com/eslint/eslint/issues/11903. For `eslint:recommended`, the criteria is that rules should only catch possible errors or disallow legacy syntax. This rule can catch TDZ errors, but is otherwise mostly stylistic.

**Comment by Slapbox** [comment]
I guess that discussion makes sense but what about a simple case like this? I'd expect this to be a warning or error by default if it occurs:

```
function example() {
	console.log(foo)
	const foo = 1;
}
```

Is there any way to implement such a minimalist version of the rule that only checks within a current function scope, or that otherwise the variable exists at least somewhere in the file? I guess that couldn't rightly be called `no-use-before-define` anymore, but it would be beneficial I think.

**Comment by mdjermanovic** [comment]
@Slapbox feel free to open a new issue with a concrete proposal, and we can definitely discuss this again (as noted in https://github.com/eslint/eslint/issues/11903#issuecomment-817387300). I generally agree that `eslint:recommended` should catch such obvious TDZ errors. The difficulty here is how to split responsibility between rules as we'd like to avoid overlapping (multiple rules reporting the same problem).
